File: packages/react-native-storybook/ios/ErrorHelper.m
#import "ErrorHelper.h"
#import "FileSystemHelper.h"

static NSString *PROTOCOL_FILENAME = @"protocol.sherlo";
static NSString *const LOG_TAG = @"ErrorHelper";

@implementation ErrorHelper

// Handles errors by logging them and writing to the protocol file
+ (void)handleError:(NSString *)errorCode error:(id)error syncDirectoryPath:(NSString *)syncDirectoryPath {
  NSString *errorDescription;
  if ([error isKindOfClass:[NSError class]]) {
    errorDescription = [(NSError *)error localizedDescription] ?: @"N/A";
  } else if ([error isKindOfClass:[NSString class]]) {
    errorDescription = (NSString *)error;
  } else {
    errorDescription = [NSString stringWithFormat:@"%@", error];
  }
  
  NSLog(@"[%@] Error occurred: %@, Error: %@", LOG_TAG, errorCode, errorDescription);

  NSString *protocolFilePath = [syncDirectoryPath stringByAppendingPathComponent:PROTOCOL_FILENAME];
  
  NSMutableDictionary *nativeErrorDict = [@{
    @"action": @"NATIVE_ERROR",
    @"errorCode": errorCode,
    @"error": errorDescription,
    @"timestamp": @((long long)([[NSDate date] timeIntervalSince1970] * 1000)),
    @"entity": @"app"
  } mutableCopy];

  // Convert the dictionary to JSON data
  NSError *jsonError;
  NSData *jsonData = [NSJSONSerialization dataWithJSONObject:nativeErrorDict options:0 error:&jsonError];
  
  if (jsonError) {
    NSLog(@"[%@] Error creating JSON: %@", LOG_TAG, jsonError.localizedDescription);
    return;
  }
  
  // Convert JSON data to base64 string
  NSString *base64ErrorData = [jsonData base64EncodedStringWithOptions:0];
  
  NSError *writeError = [FileSystemHelper appendFile:protocolFilePath contents:base64ErrorData];
  if (writeError) {
    NSLog(@"[%@] Error writing to error file: %@", LOG_TAG, writeError.localizedDescription);
  }
}

/**
 * Instance method that delegates to the class method for handling errors.
 */
- (void)handleError:(NSString *)errorCode error:(id)error syncDirectoryPath:(NSString *)syncDirectoryPath {
  [ErrorHelper handleError:errorCode error:error syncDirectoryPath:syncDirectoryPath];
}

@end 

File: packages/react-native-storybook/android/src/main/java/io/sherlo/storybookreactnative/ErrorHelper.java
package io.sherlo.storybookreactnative;

import android.util.Base64;
import android.util.Log;
import org.json.JSONObject;
import java.util.HashMap;
import java.util.Map;

/**
 * Helper class for handling and logging errors in the Sherlo module.
 * Centralizes error handling and protocol file writing.
 */
public class ErrorHelper {
    public static final String TAG = "ErrorHelper";
    private static final String PROTOCOL_FILENAME = "protocol.sherlo";

    private final FileSystemHelper fileSystemHelper;
    private final String syncDirectoryPath;

    public ErrorHelper(FileSystemHelper fileSystemHelper, String syncDirectoryPath) {
        this.fileSystemHelper = fileSystemHelper;
        this.syncDirectoryPath = syncDirectoryPath;
    }

    /**
     * Handles errors by logging them and writing to the protocol file.
     * 
     * @param errorCode The error code
     * @param errorMessage The error message
     */
    public void handleError(String errorCode, String errorMessage) {
        Log.e(TAG, "Error occurred: " + errorCode + ", Error: " + errorMessage);

        String protocolFilePath = this.syncDirectoryPath + "/" + PROTOCOL_FILENAME;

        // Create the new JSON object with the specified properties
        Map<String, Object> nativeErrorMap = new HashMap<>();
        nativeErrorMap.put("action", "NATIVE_ERROR");
        nativeErrorMap.put("errorCode", errorCode);
        nativeErrorMap.put("error", errorMessage);
        nativeErrorMap.put("timestamp", System.currentTimeMillis());
        nativeErrorMap.put("entity", "app");

        try {
            // Convert the map to JSON string
            String jsonString = new JSONObject(nativeErrorMap).toString();

            // Convert JSON string to base64
            String base64ErrorData = Base64.encodeToString(jsonString.getBytes("UTF-8"),
                    Base64.NO_WRAP);

            // Append the base64 encoded error data to the protocol file
            fileSystemHelper.appendFile(protocolFilePath, base64ErrorData);
        } catch (Exception e) {
            Log.e(TAG, "Error writing to error file: " + e.getMessage());
        }
    }

    /**
     * Handles an exception by extracting its message and passing it to handleError.
     * 
     * @param errorCode The error code
     * @param exception The exception to handle
     */
    public void handleException(String errorCode, Exception exception) {
        handleError(errorCode, exception.getMessage());
    }
} 